### DISCUSSION TITLE
How to update a vertex?

### DISCUSSION THREAD

**Discussion by bCamba** [Gremlinq/ExRam.Gremlinq#709] [Q&A] [answered]
I am getting the following error when trying to do a vertex update (I am using CosmosDb provider if that makes a difference):

> ExceptionType : GraphSyntaxException
ExceptionMessage : 
	Gremlin query syntax error: Partition key property of a vertex is readonly. Please consider deleting and recreating the vertex with the desired partition key. Partition key property on an edge  is considered system property and can't neither be set or updated.

Sample code:

```
    [HttpGet(Name = "GetWeatherForecast")]
    public async Task<Document> Get()
    {
        var test = new Document("test", 10);
        await _g.AddV(test)
            .FirstAsync();

        var documentFromDb = await _g.V<Document>(test.Id).FirstAsync();

        var newDocument = documentFromDb with
        {
            Number = 100
        };

        await _g.V(newDocument.Id)
            .Update(newDocument);
        
        return test;
    }

    public record Document : BaseAuditableEntity
    {
        public Document(string name, int number) : base(Guid.NewGuid())
        {
            Name = name;
            Number = number;
        }
    
        public string Name { get; init; }
    
        public int Number { get; init; }
    };

    public abstract record BaseAuditableEntity : BaseEntity
    {
        public BaseAuditableEntity(Guid createdBy, Guid modifiedBy, DateTime createdTime)
        {
            CreatedBy = createdBy;
            ModifiedBy = modifiedBy;
            CreatedTime = createdTime;
            LastModified = DateTime.UtcNow;
        }
    
        public BaseAuditableEntity(Guid userId)
        {
            CreatedBy = userId;
            ModifiedBy = userId;
    
            var creationTime = DateTime.UtcNow;
    
            CreatedTime = creationTime;
            LastModified = creationTime;
        }
    
        public DateTime CreatedTime { get; init; }
        public DateTime LastModified { get; init; }
        public Guid CreatedBy { get; init; }
        public Guid ModifiedBy { get; init; }
    }

    public abstract record BaseEntity
    {
        public Guid Id { get; init; } = Guid.NewGuid();
    }
```

**Comment by danielcweber** [comment, marked as answer]
Unfortunately you omitted how you setup Gremlinq, but make sure you exclude the partition key from updates (either [by using this](https://github.com/ExRam/ExRam.Gremlinq.Samples/blob/e1f7393f534cc1efd6453040aba352d4cd0ae1bb/ExRam.Gremlinq.Samples.AspNet/Startup.cs#L25) or [this](https://github.com/ExRam/ExRam.Gremlinq.Samples/blob/e1f7393f534cc1efd6453040aba352d4cd0ae1bb/ExRam.Gremlinq.Samples.Console/Program.cs#L35))

**Comment by bCamba** [comment]
I tried using [this](https://github.com/ExRam/ExRam.Gremlinq.Samples/blob/e1f7393f534cc1efd6453040aba352d4cd0ae1bb/ExRam.Gremlinq.Samples.AspNet/Startup.cs#L25) as you recommended and it solved this problem. However is there a particular reason why using the same base classes for vertices and edges leads to this error?

> System.ArgumentException: Wholechain.BFF.Api.Domain.Common.BaseAuditableEntity may not be in the inheritance hierarchy of Wholechain.BFF.Api.Domain.Common.BaseAuditableEntity.

**Reply by danielcweber** [reply]
That case is just an edge case of the "X may not be in the inheritance hierarchy of Y", which is reasonable. I would however, strongly advise against using the same base classes for vertices and edges.

**Reply by bCamba** [reply]
Yes after a little bit more digging I understand better know. Yes it makes sense for them not to have the same base classes. Thank you
